modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/InlineUtils.java

```diff
@@ -0,0 +1,147 @@
+import static org.apache.ignite.internal.util.Constants.KiB;
+
+import java.util.List;
+import org.apache.ignite.internal.binarytuple.BinaryTupleCommon;
+import org.apache.ignite.internal.pagememory.tree.io.BplusInnerIo;
+import org.apache.ignite.internal.pagememory.tree.io.BplusLeafIo;
+import org.apache.ignite.internal.schema.BinaryTuple;
+import org.apache.ignite.internal.schema.NativeType;
+import org.apache.ignite.internal.schema.NativeTypeSpec;
+import org.apache.ignite.internal.schema.VarlenNativeType;
+import org.apache.ignite.internal.storage.index.IndexDescriptor;
+import org.apache.ignite.internal.storage.index.IndexDescriptor.ColumnDescriptor;
+
+/**
+ * Helper class for index inlining.
+ */
+public class InlineUtils {
+    /** Maximum inline size for a {@link BinaryTuple}, in bytes. */
+    public static final int MAX_BINARY_TUPLE_INLINE_SIZE = 2 * KiB;
+
+    /**
+     * Calculates inline size for column.
+     *
+     * @param nativeType Column type.
+     * @param maxVarlenInlineSize Maximum inline size of variable length column in bytes.
+     * @return Inline size in bytes.
+     */
+    static int inlineSize(NativeType nativeType, int maxVarlenInlineSize) {
+        NativeTypeSpec spec = nativeType.spec();
+
+        if (spec.fixedLength()) {
+            return nativeType.sizeInBytes();
+        }
+
+        // Variable length columns.
+
+        switch (spec) {
+            case STRING:
+            case BYTES:
+                return Math.min(maxVarlenInlineSize, ((VarlenNativeType) nativeType).length());
+
+            case DECIMAL:
+                // It can be converted to float for greater or lesser comparisons.
+                return Float.BYTES;
+
+            case NUMBER:
+                // It can be converted to int for greater or lesser comparisons.
+                return Integer.BYTES;
+
+            default:
+                throw new IllegalArgumentException("Unknown type " + spec);
+        }
+    }
+
+    /**
+     * Calculates inline size for {@link BinaryTuple}, given its format.
+     *
+     * @param indexDescriptor Index descriptor.
+     * @param maxVarlenInlineSize Maximum inline size of variable length column in bytes.
+     * @return Inline size in bytes, not more than the {@link #MAX_BINARY_TUPLE_INLINE_SIZE}.
+     */
+    static int binaryTupleInlineSize(IndexDescriptor indexDescriptor, int maxVarlenInlineSize) {
+        List<? extends ColumnDescriptor> columns = indexDescriptor.columns();
+
+        boolean hasNullColumns = columns.stream().anyMatch(ColumnDescriptor::nullable);
+
+        // 2-byte size class covers most cases.
+        int sizeClass = 2;
+
+        if (columns.stream().allMatch(c -> c.type().spec().fixedLength())) {
+            int totalFixedColumnSize = columns.stream().mapToInt(c -> c.type().sizeInBytes()).sum();
+
+            sizeClass = Math.min(sizeClass, BinaryTupleCommon.flagsToEntrySize(BinaryTupleCommon.valueSizeToFlags(totalFixedColumnSize)));
```

Comment:
This line is so messy. We can use static imports or extract some variables, what do you think?

Reply:
Also, it's a double conversion, not very intuitive. Can we have "valueSizeToEntrySize"?
BTW, entry size if better than "class"

Reply:
Fix it.